### PR TITLE
fix(bonds): deployment guards + dust-mint defence

### PR DESCRIPTION
**Pull request by tiero**
Follow-up to the merged #37. Addresses the three actionable items from arkanaai's automated review ([#pullrequestreview-4404098591](https://github.com/arkade-os/compiler/pull/37#pullrequestreview-4404098591)) — one P1 + two P2. No protocol-breaking changes; all new logic is at the deployment-safety layer.

### P1 — `auctionWindow > 0` deployment guard (`issue` + `rollIn`)

Without it, `auctionWindow = 0` collapses the auction time gate `tx.time >= maturity AND tx.time < maturity` to the empty set, so no defaulted vault can EVER be settled via `acceptAuction`. `mintedAmount` stays in `totalDebitOutstanding` forever and credit holders absorb 100% of every default. Same class of deployment invariant as the existing `initRatioBps > liqThresholdBps` and `liqThresholdBps > 0` checks.

### P2 — `auctionDiscountBps` validation at deployment (`issue` + `rollIn`)

Previously checked only at runtime in `liquidate()` and `acceptAuction()`. An out-of-range discount bricks both settlement paths, leaving the pool unsettleable — yet a vault could still be issued. Moved `require(auctionDiscountBps >= 0 && < 10000)` into the issuance deployment-safety block; runtime checks retained as defense-in-depth.

### P2 — Dust-issuance ceiling division on origination floor (`issue` + `rollIn`)

`required = amount * initRatioBps / 10000` floors to 1 for `amount=1, initRatioBps=14999`, letting 1 sat of collateral mint 1 credit + 1 debit. An attacker could flood the auction window with thousands of dust defaults that are unprofitable for auctioneers to settle (gas > profit), leaving them as permanent deadweight in `totalDebitOutstanding`. Replaced with ceiling:

```
required = (amount * initRatioBps + 9999) / 10000
```

Ceiling rounds UP so `1 * 14999 / 10000` → 2, breaking the unit-boundary mint hole. A configurable `minAmount` / `minCollateral` floor remains follow-up R1 (the ceiling alone is necessary but not sufficient for very-small-but-non-dust amounts).

## Tests

| Test | Change |
|---|---|
| `test_issue_enforces_liq_threshold_invariants` | Renamed → `test_issue_enforces_deployment_invariants`; OP_GREATERTHAN count 5 → 6 (added `auctionWindow > 0`); new `OP_GREATERTHANOREQUAL64 >= 1` floor for the `auctionDiscountBps` deployment check |
| `test_roll_pair_enforces_both_threshold_invariants` | Renamed → `test_roll_pair_enforces_all_deployment_invariants`; same updates |
| `test_issue_uses_ceiling_division_on_required_collateral` | **NEW** — asserts the literal `9999` token appears in both `issue` and `rollIn` ASM. A regression to floor (or to a different bias) trips it |

255 → 256 tests, all green.

## Docs

- `docs/bonds.md` §Enforced deployment invariants expanded from 2 → 4 and now applies to `rollIn` as well as `issue`.
- New paragraph on the ceiling-division origination floor.
- Follow-up R2 marked **WIRED** with reference to the new regression test; R1 kept as a separate open item (configurable minimums).

## Out of scope here

- **arkanaai P2.4** — `examples/lending/loan_vault.ark` still uses `>=` burn checks. That's a different contract family with a different trust model (keeper-signed vs permissionless); deferring to a separate PR.
- **arkanaai P3 observations** — the four P3 items (#5-#8) are informational and require no action.

## Test plan

- [x] `cargo test` — 256/256 pass
- [x] `cargo fmt --check` — clean
- [x] ASM verification: `<auctionWindow>` placeholder + literal `9999` present in both `issue` and `rollIn` server variants
- [x] OP_GREATERTHAN count = 6 in both `issue` and `rollIn` (was 5)

https://claude.ai/code/session_013U1kfFDEPH91g5VnPTmKD4

---
_Generated by [Claude Code](https://claude.ai/code/session_013U1kfFDEPH91g5VnPTmKD4)_

## Summary by CodeRabbit

* **Documentation**
  * Updated bond pool documentation with expanded details on issuance requirements and collateral calculation rules.

* **Bug Fixes**
  * Improved collateral calculation accuracy to handle rounding edge cases.
  * Added parameter validation for bond pool operations.

* **Tests**
  * Enhanced coverage for bond pool safety invariant validation.

### CODE REVIEW

**Review by ghost — CHANGES_REQUESTED** [review]
## arkanaai review — arkade-os/compiler#38

Reviewed: `examples/bonds/repayment_pool.ark`, `tests/repayment_pool_test.rs`, `docs/bonds.md`. Read full files, full diff, git history, cross-repo grep.

### Verdict: Code is correct. Requesting human sign-off (protocol-critical).

All three changes (P1 auctionWindow guard, P2 auctionDiscountBps guard, P2 ceiling division) are sound. The deployment guards close real deployment-configuration attack surfaces, the ceiling division closes a real dust-mint vector. No cross-repo consumers; no API changes. Docs updated. Tests added.

I'm requesting changes **solely** because this is protocol-critical code (collateralisation math + deployment guards on a bond market) and our policy requires human sign-off. Once a human reviewer confirms, this is good to merge.

---

### ✅ What's correct

**1. `auctionWindow > 0` guard** (`repayment_pool.ark:150,355`) — Correct. Without it, `tx.time >= maturity AND tx.time < maturity + 0` is the empty set; no vault can ever be settled via `acceptAuction`. Permanent deadweight in `totalDebitOutstanding`. Both `issue` and `rollIn` enforce it.

**2. `auctionDiscountBps ∈ [0, 10000)` guard** (`repayment_pool.ark:151-152,356-357`) — Correct. A negative discount makes `payRate > 10000`, demanding >100% of par from auctioneers — economically unsettleable. A discount ≥ 10000 makes `payRate ≤ 0` — zero or negative USDT enters the pool. Both failure modes brick the settlement layer. Runtime checks in `liquidate`/`acceptAuction` retained as defence-in-depth.

**3. Ceiling division** (`repayment_pool.ark:175,371`) — Correct. `(amount * initRatioBps + 9999) / 10000` is the standard ceiling-division idiom for integer arithmetic. Verified edge cases:
- `amount=1, initRatioBps=14999` → `(14999+9999)/10000 = 2` (was 1 with floor — the hole)
- `amount=1, initRatioBps=9999` → `(9999+9999)/10000 = 1` (was 0 with floor — even worse hole)
- `amount=1, initRatioBps=10000` → `(10000+9999)/10000 = 1` (unchanged, exact multiple)
- Overflow: `+9999` is negligible vs the existing `amount * initRatioBps` product; `OP_MUL64` already fails closed on overflow.

**4. Test coverage** — All three changes have regression tests:
- `test_issue_enforces_deployment_invariants`: exact OP_GREATERTHAN count = 6 (was 5). Removal of any single guard fails the test.
- `test_roll_pair_enforces_all_deployment_invariants`: mirrors issue for rollIn.
- `test_issue_uses_ceiling_division_on_required_collateral`: asserts literal `9999` in ASM for both `issue` and `rollIn`.

**5. Cross-repo impact** — None. All references to `RepaymentPool`, `auctionDiscountBps`, `auctionWindow`, `initRatioBps`, `liqThresholdBps` are contained within the compiler repo (examples, tests, docs). No downstream SDK or consumer breakage.

---

### ⚠️ Observations (non-blocking)

**O1. Weak `OP_GREATERTHANOREQUAL64 >= 1` assertion** — `tests/repayment_pool_test.rs:276-281,694-698`

The `auctionDiscountBps >= 0` deployment check is tested via `assert!(gte >= 1)` on the OP_GREATERTHANOREQUAL64 opcode count. The test comment acknowledges this is weaker than the exact-count pattern used for OP_GREATERTHAN. In practice, the `issue` function emits *many* OP_GREATERTHANOREQUAL64 instructions for asset-amount `>=` checks (e.g. `tx.outputs[0].assets.lookup(usdtAssetId) >= usdtBalance`), so the `>= 1` floor would still pass even if the `auctionDiscountBps >= 0` check were removed. The upper-bound check (`< 10000`) is separately guarded by the OP_LESSTHAN count in other tests, but the lower-bound check lacks a strong regression guard.

Not blocking — the deployment guard is correct in the source, and removing it would require deliberate action. But worth noting that this is the one invariant without a tight opcode-count regression test. A future option: count the exact OP_GREATERTHANOREQUAL64 total and pin it.

**O2. `healthFloor` in `liquidate` still uses floor division** — `repayment_pool.ark:463`

`healthFloor = mintedAmount * liqThresholdBps / 10000` floors. For dust-sized `mintedAmount` values, this floors to 0, making `collateralValue < 0` impossible — the vault becomes unliquidatable. The ceiling division on origination raises the collateral bar, but a vault with `mintedAmount=1` and `liqThresholdBps=9999` still has `healthFloor=0` and can never be margin-called.

This is acknowledged in the PR as follow-up R1 (configurable minimums). Floor division on `healthFloor` is actually the *conservative* direction (protects borrowers from premature liquidation), so it's not a bug — it's just the residual dust problem that R1 will close. No action needed here.

---

### 📋 Protocol-critical flag

This PR modifies:
- **Collateralisation math** (ceiling division changes which vaults can be created)
- **Deployment guards** (changes which pool configurations are valid)

Per policy: **requires explicit human sign-off before merge**, even though the code is correct.

🤖 Generated with [Claude Code](https://claude.ai/code)

**Review by ghost — CHANGES_REQUESTED** [review]
## arkanaai review — arkade-os/compiler#38

Reviewed: `examples/bonds/repayment_pool.ark`, `tests/repayment_pool_test.rs`, `docs/bonds.md`. Read full files, full diff, git history, cross-repo grep.

### Verdict: Code is correct. Requesting human sign-off (protocol-critical).

All three changes (P1 auctionWindow guard, P2 auctionDiscountBps guard, P2 ceiling division) are sound. The deployment guards close real deployment-configuration attack surfaces, the ceiling division closes a real dust-mint vector. No cross-repo consumers; no API changes. Docs updated. Tests added.

I'm requesting changes **solely** because this is protocol-critical code (collateralisation math + deployment guards on a bond market) and our policy requires human sign-off. Once a human reviewer confirms, this is good to merge.

---

### ✅ What's correct

**1. \`auctionWindow > 0\` guard** (\`repayment_pool.ark:150,355\`) — Correct. Without it, \`tx.time >= maturity AND tx.time < maturity + 0\` is the empty set; no vault can ever be settled via \`acceptAuction\`. Permanent deadweight in \`totalDebitOutstanding\`. Both \`issue\` and \`rollIn\` enforce it.

**2. \`auctionDiscountBps ∈ [0, 10000)\` guard** (\`repayment_pool.ark:151-152,356-357\`) — Correct. A negative discount makes \`payRate > 10000\`, demanding >100% of par from auctioneers — economically unsettleable. A discount ≥ 10000 makes \`payRate ≤ 0\` — zero or negative USDT enters the pool. Both failure modes brick the settlement layer. Runtime checks in \`liquidate\`/\`acceptAuction\` retained as defence-in-depth.

**3. Ceiling division** (\`repayment_pool.ark:175,371\`) — Correct. \`(amount * initRatioBps + 9999) / 10000\` is the standard ceiling-division idiom for integer arithmetic. Verified edge cases:
- \`amount=1, initRatioBps=14999\` → \`(14999+9999)/10000 = 2\` (was 1 with floor — the hole)
- \`amount=1, initRatioBps=9999\` → \`(9999+9999)/10000 = 1\` (was 0 with floor — even worse hole)
- \`amount=1, initRatioBps=10000\` → \`(10000+9999)/10000 = 1\` (unchanged, exact multiple)
- Overflow: \`+9999\` is negligible vs the existing \`amount * initRatioBps\` product; \`OP_MUL64\` already fails closed on overflow.

**4. Test coverage** — All three changes have regression tests:
- \`test_issue_enforces_deployment_invariants\`: exact OP_GREATERTHAN count = 6 (was 5). Removal of any single guard fails the test.
- \`test_roll_pair_enforces_all_deployment_invariants\`: mirrors issue for rollIn.
- \`test_issue_uses_ceiling_division_on_required_collateral\`: asserts literal \`9999\` in ASM for both \`issue\` and \`rollIn\`.

**5. Cross-repo impact** — None. All references to \`RepaymentPool\`, \`auctionDiscountBps\`, \`auctionWindow\`, \`initRatioBps\`, \`liqThresholdBps\` are contained within the compiler repo (examples, tests, docs). No downstream SDK or consumer breakage.

---

### ⚠️ Observations (non-blocking)

**O1. Weak \`OP_GREATERTHANOREQUAL64 >= 1\` assertion** — \`tests/repayment_pool_test.rs:276-281,694-698\`

The \`auctionDiscountBps >= 0\` deployment check is tested via \`assert!(gte >= 1)\` on the OP_GREATERTHANOREQUAL64 opcode count. The test comment acknowledges this is weaker than the exact-count pattern used for OP_GREATERTHAN. In practice, the \`issue\` function emits *many* OP_GREATERTHANOREQUAL64 instructions for asset-amount \`>=\` checks (e.g. \`tx.outputs[0].assets.lookup(usdtAssetId) >= usdtBalance\`), so the \`>= 1\` floor would still pass even if the \`auctionDiscountBps >= 0\` check were removed. The upper-bound check (\`< 10000\`) is separately guarded by the OP_LESSTHAN count in other tests, but the lower-bound check lacks a strong regression guard.

Not blocking — the deployment guard is correct in the source, and removing it would require deliberate action. But worth noting that this is the one invariant without a tight opcode-count regression test. A future option: count the exact OP_GREATERTHANOREQUAL64 total and pin it.

**O2. \`healthFloor\` in \`liquidate\` still uses floor division** — \`repayment_pool.ark:463\`

\`healthFloor = mintedAmount * liqThresholdBps / 10000\` floors. For dust-sized \`mintedAmount\` values, this floors to 0, making \`collateralValue < 0\` impossible — the vault becomes unliquidatable. The ceiling division on origination raises the collateral bar, but a vault with \`mintedAmount=1\` and \`liqThresholdBps=9999\` still has \`healthFloor=0\` and can never be margin-called.

This is acknowledged in the PR as follow-up R1 (configurable minimums). Floor division on \`healthFloor\` is actually the *conservative* direction (protects borrowers from premature liquidation), so it's not a bug — it's just the residual dust problem that R1 will close. No action needed here.

---

### 📋 Protocol-critical flag

This PR modifies:
- **Collateralisation math** (ceiling division changes which vaults can be created)
- **Deployment guards** (changes which pool configurations are valid)

Per policy: **requires explicit human sign-off before merge**, even though the code is correct.

🤖 Generated with [Claude Code](https://claude.ai/code)

File: tests/repayment_pool_test.rs [hunk 1]
```diff
@@ -237,28 +237,75 @@ fn test_issue_is_oracle_priced_and_dual_mints() {
...
+    // `output.assets.lookup(...) >= N` use the same opcode, but those fire
+    // AFTER the deployment checks and aren't counted as deployment guards.
+    // Floor at 1 captures the deployment check without coupling to the
+    // exact count of asset-amount checks.)
+    let gte = opcode_count(&output, "issue", "OP_GREATERTHANOREQUAL64");
+    assert!(
+        gte >= 1,
+        "issue must carry auctionDiscountBps >= 0 (OP_GREATERTHANOREQUAL64, \
+         found {gte} total — at least one must be the discount-bound check)"
+    );
```
[the remark this answers is not in the document]
Comment (tiero) [reply tests/repayment_pool_test.rs]: Addressed in de7e4fe.

Replaced both sites (`tests/repayment_pool_test.rs:281` and the parallel rollIn assertion) with a targeted 3-token window check that anchors specifically to the `<auctionDiscountBps>` guard:

```rust
contains_window_3(output, "issue",
                  "<auctionDiscountBps>", "OP_GREATERTHANOREQUAL", "0")
```

plus the symmetric `< 10000` upper-bound check. The helper accepts any permutation of the three tokens within a 3-token sliding window — the emitter's current display order for the discount comparison is `<auctionDiscountBps> OP_GREATERTHANOREQUAL 0`, which isn't strict postfix, so the set-equality check is robust to that emitter quirk and to future reordering.

**Verified by regression injection**: deleting `require(auctionDiscountBps >= 0, ...)` from `issue()` makes `test_issue_enforces_deployment_invariants` fail with the message *"issue must carry the \`auctionDiscountBps >= 0\` deployment guard (expected window of \<auctionDiscountBps\>, OP_GREATERTHANOREQUAL, 0 within 3 consecutive ASM tokens of \`issue\`)"*; restoring it passes. Same for `rollIn`.

256/256 tests still pass.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by ghost — APPROVED** [review]
## arkanaai review — arkade-os/compiler#38 (incremental, commit de7e4fe)

**Scope:** Only the second commit (`de7e4fed` — "anchor auctionDiscountBps assertions to the specific guard"). Contract code (`repayment_pool.ark`) and docs are unchanged from my prior review.

### ✅ O1 resolved — test assertions are now tight

The weak `gte >= 1` floor on `OP_GREATERTHANOREQUAL64` (my O1 from the first review, independently flagged by CodeRabbit) is replaced with a targeted `contains_window_3` check that anchors to the **specific** `<auctionDiscountBps>` operand:

- `tests/repayment_pool_test.rs` — `test_issue_enforces_deployment_invariants`: asserts `<auctionDiscountBps>`, `OP_GREATERTHANOREQUAL`, `0` appear as a 3-token window in `issue` ASM. Also asserts the `< 10000` upper bound (`<auctionDiscountBps>`, `OP_LESSTHAN`, `10000`).
- Same assertions mirrored in `test_roll_pair_enforces_all_deployment_invariants` for `rollIn`.

**`contains_window_3` helper** — Correct. Uses `BTreeSet` set-equality over a sliding window of 3 tokens, so order within the window doesn't matter (robust to emitter reordering). All call sites use three distinct tokens, so `BTreeSet` dedup is a non-issue. The 3-token adjacency constraint makes false positives on stray `0` or `10000` literals practically impossible.

Commit message confirms regression injection was verified: deleting the `require(auctionDiscountBps >= 0, ...)` guard from `issue()` fails the test with the expected message.

### 📋 Protocol-critical flag — still applies

My prior approval of the contract code stands. The protocol-critical human sign-off requirement from my earlier reviews remains: this PR modifies collateralisation math and deployment guards on a bond market. **A human must approve before merge.**

Approving from my side — all code and tests are correct. The ball is with a human reviewer.

🤖 Generated with [Claude Code](https://claude.ai/code)
